From: Christian Kastner <ckk@kvr.at>
Date: Fri, 15 Jan 2016 20:35:09 +0100
Subject: Suppress crontab header by default

Change the default behaviour including a warning header when editing a crontab
to not doing so. This makes `crontab -l | crontab -` idempotent.

Adding the header can be fored by exporting CRONTAB_NOHEADER=N.

Contributed by Steve Greenland <stevegr@debian.org>.

Forwarded: no
Last-Update: 2016-01-15
Index: cron/crontab.c
===================================================================
--- cron.orig/crontab.c
+++ cron/crontab.c
@@ -292,6 +292,8 @@ list_cmd() {
 	char	n[MAX_FNAME];
 	FILE	*f;
 	int	ch;
+	int	x;
+	char	*ctnh;
 
 	log_it(RealUser, Pid, "LIST", User);
 	(void) snprintf(n, MAX_FNAME, CRON_TAB(User));
@@ -307,6 +309,33 @@ list_cmd() {
 	/* file is open. copy to stdout, close.
 	 */
 	Set_LineNum(1)
+
+	/* Don't list header lines unless CRONTAB_NOHEADER is 'N'.
+	 *
+	 * ignore the top few comments since we probably put them there.
+	 */
+	if (!(ctnh = getenv("CRONTAB_NOHEADER")) || toupper(*ctnh) != 'N') {
+		for (x = 0;  x < NHEADER_LINES;  x++) {
+			ch = get_char(f);
+			if (EOF == ch) {
+				break;
+			}
+
+			if ('#' != ch) {
+				putchar(ch);
+				break;
+			}
+			while (EOF != (ch = get_char(f))) {
+				if (ch == '\n') {
+					break;
+				}
+			}
+			if (EOF == ch) {
+				break;
+			}
+		}
+	}
+
 	while (EOF != (ch = get_char(f)))
 		putchar(ch);
 	fclose(f);
Index: cron/crontab.1
===================================================================
--- cron.orig/crontab.1
+++ cron/crontab.1
@@ -82,7 +82,10 @@ named file or standard input if the pseu
 .PP
 The
 .I \-l
-option causes the current crontab to be displayed on standard output.
+option causes the current crontab to be displayed on standard output.  See
+the note under
+.B DEBIAN SPECIFIC
+below.
 .PP
 The
 .I \-r
@@ -97,6 +100,26 @@ from the editor, the modified crontab wi
 neither of the environment variables is defined, then the
 default editor /usr/bin/editor is used.
 .PP
+.SH DEBIAN SPECIFIC
+The "out-of-the-box" behaviour for
+.I crontab \-l
+is to display the three line "DO NOT EDIT THIS FILE" header
+that is placed at the
+beginning of the crontab when it is installed.  The problem is that
+it makes the sequence
+.PP
+crontab \-l | crontab \-
+.PP
+non-idempotent \(em you keep adding copies of the header.  This causes
+pain to scripts that use sed to edit a crontab.  Therefore, the default
+behaviour of the
+.B \-l
+option has been changed to not output such header.  You may obtain the
+original behaviour by setting the environment variable
+.B CRONTAB_NOHEADER
+to 'N', which will cause the
+.I crontab \-l
+command to emit the extraneous header.
 .SH "SEE ALSO"
 crontab(5), cron(8)
 .SH FILES
